### PR TITLE
CID update to RFC 9146

### PR DESCRIPTION
**Pull request by hannestschofenig**
## Description

The DTLS 1.2 CID specification is finally close to publication and will be released as RFC 9146. This PR updates the implementation to match the RFC content.

Here is the CID RFC in the RFC Editor Queue: https://www.rfc-editor.org/authors/rfc9146.html

There are now two compile-time options in this PR, namely 
 * MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY  (for the CID draft version draft-ietf-tls-dtls-connection-id-05), and 
 * MBEDTLS_SSL_DTLS_CONNECTION_ID (for the RFC 9146 version). 
These two options can only be used mutuallly exclusive. 

## Status

**IN DEVELOPMENT**

## Requires Backporting

This PR does not require backporting. 

## Migrations

There is NO API change. 

## Additional comments

IMPORTANT: For those who used the temporary CID implementation in Mbed TLS please note that there were late changes to the specification that are backwards-compatibility-breaking. Two aspects have changed: 

* IANA assigned a different value to the CID extension. The Mbed TLS implementation so far used an experimental value. 
* The MAC computation and the additional data computation changed. 

## Todos

- [] Tests
- [x] Documentation
- [ ] Changelog updated

## Steps to test or reproduce

I have tested ssl_client2 against Californium.

### CODE REVIEW

File: library/ssl_msg.c [hunk 1]
```diff
@@ -376,31 +376,75 @@ static int ssl_parse_inner_plaintext( unsigned char const *content,
...
+ * factors, namely
+ *
+ * 1) CID functionality disabled
+ *
+ * size = 13 bytes
+ *
+ * 2) CID functionality based on RFC 9146 enabled
+ *
+ * size = 8 + 1 +1 +1 + 2 + 2 + 6 + n + 2
+ *      = 13 + 1 + 8 + 1 = 23 bytes
```
Comment (boaks) [inline library/ssl_msg.c:392]: missing "+ n" on both sides of the "=".
And definition n := CID-length

File: library/ssl_msg.c [hunk 2]
```diff
@@ -435,25 +482,49 @@ static void ssl_extract_add_data_from_record( unsigned char* add_data,
...
+
+            // type
+            *cur = rec->type;
+            cur++;
+
+            // cid_length
+            *cur = rec->cid_len;
+            cur++;
+        }
+#else
```
[the remark this answers is not in the document]
Comment (hannestschofenig) [reply library/ssl_msg.c:558]: I put all cases into the ssl_extract_add_data_from_record() function.

**Comment by hannestschofenig** [comment]
Travis fails at a point where it should fail. For some reason, it defines MBEDTLS_SSL_DTLS_CONNECTION_ID  and MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY  at the same time. This is not allowed, which will lead to this error:

 #error "MBEDTLS_SSL_DTLS_CONNECTION_ID and MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY cannot be defined simultaneously"

**Comment by mpg** [comment]
Lots of build in the CI try to enable everything for maximal coverage. I think you want to add `MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY` to the `EXCLUDE_FROM_FULL` list in `scripts/config.py`.

And since `MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY` is disabled both in the default config and in the "full" config, we'll want a dedicated component in `tests/scripts/all.sh` that builds with that option enabled and runs a few tests. I think something similar to the existing `component_test_no_renegotiation` would be good.

**Review by gergelynagymate — CHANGES_REQUESTED** [review]
Hi,
I suggest this bug fix might be incorrect. There should stay the #else #endif, but the epoch and sequence number should be also appended just above in an else branch (when rec->cid_len==0). 
The present solution appends the epoch and sequence number twice resulting incorrect MAC calculation:

additional_data

Thanks for considering my feedback. 
Best regards, Gergely

**Comment by boaks** [comment]
Implementing the different variants in [Eclipse/Californium](
https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java#L488-L507), I struggled and decided to split it in different function. Maybe, that's also an idea to go.

**Comment by boaks** [comment]
I guess, the `#endif` in line 558, needs to be a `#else` and a `#endif` must be added between line 561 and 562.
 
[wrong `#endif`](https://github.com/ARMmbed/mbedtls/blob/3221ee21c6da4047e8629241ce706c8a2b78b0b5/library/ssl_msg.c#L558-L562)

**Comment by gergelynagymate** [comment]
> 
> 
> I guess, the `#endif` in line 558, needs to be a `#else` and a `#endif` must be added between line 561 and 562.
> 
> [wrong `#endif`](https://github.com/ARMmbed/mbedtls/blob/3221ee21c6da4047e8629241ce706c8a2b78b0b5/library/ssl_msg.c#L558-L562)
> 

Yes, as it was before the commit, but the epoch and sequence should be also added when MBEDTLS_SSL_DTLS_CONNECTION_ID  is defined but rec->cid_len is 0, as the case with messages coming from the server which does not include CID. 
In this case RFC6347 and RFC7366 applies, am I correct?

    #if defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
            if (rec->cid_len != 0)
            {
                // seq_num_placeholder
                memcpy(cur, seq_num_placeholder, sizeof(seq_num_placeholder));
                cur += sizeof(seq_num_placeholder);
    
                // tls12_cid type
                *cur = rec->type;
                cur++;
    
                // cid_length
                *cur = rec->cid_len;
                cur++;
            } else {
                // epoch + sequence number
                memcpy( cur, rec->ctr, sizeof( rec->ctr ) );
                cur += sizeof( rec->ctr );
            }
    #else
          // epoch + sequence number
            memcpy( cur, rec->ctr, sizeof( rec->ctr ) );
            cur += sizeof( rec->ctr );
    #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */

This works for me both with MBEDTLS_SSL_DTLS_CONNECTION_ID defined or not.

**Comment by boaks** [comment]
I would still feel much more comfortable, if the "extract" in `ssl_extract_add_data_from_record` is split into different functions for each variant, `RFC6347`, `RFC9146` and `RFC9146 pre/deprecated`.  Using the effective cid length and NONE, CID, DEPRECATED_CID macros then in the `ssl_extract_add_data_from_record` to select the variant's function.

**Comment by boaks** [comment]
Thumbs up!

[mbedtls_cd_54.pcapng.gz](https://github.com/ARMmbed/mbedtls/files/7884097/mbedtls_cd_54.pcapng.gz)

Ensure your wireshark supports the final hello extension 54, otherwise it displays many "malformed records" ;-) (version 3.7. 0).

File: include/mbedtls/mbedtls_config.h [hunk 3]
```diff
@@ -1335,6 +1329,28 @@
...
+/**
+ * \def MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY
+ *
+ * Enable support for the DTLS Connection ID extension utilizing 
+ * the version published in draft-ietf-tls-dtls-connection-id-05,
+ * https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-05
+ *
+ * Support for this functionality is included for backwards-
+ * compatibility reasons. Note: The legacy version and the version
+ * published in RFC 9146 cannot be enabled at the same time.
```
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/mbedtls_config.h:1341]: Also, we need to add a component in `tests/scripts/all.sh` that builds with the legacy config and executes the tests (`make test` and `tests/ssl-opt.sh -f'Connection ID'` and `tests/ssl-opt.sh -f'CID').

By the way, I think existing tests in `tests/suites/test_suite_ssl.*` and `tests/ssl-op.sh` that depend on `MBEDTLS_SSL_DTLS_CONNECTION_ID` should probably now depend on that or the legacy version... which would require special treatement in `ssl-opt.sh` as it doesn't have `requires_one_of_these_enabled`.

I'm starting to wonder if the strategy for the options is right. What if we had instead:
- `MBEDTLS_SSL_DTLS_CONNECTION_ID` that controls whether _some_ version of C-ID is enabled
- `MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT` that could be set to `0` (default) for the standard version, `1` for the legacy draft version, and perhaps in the future `2` for both (negotiated at runtime) if ever implement that.

That would avoid changing all the places that currently depend on `MBEDTLS_SSL_DTLS_CONNECTION_ID`, unless I'm mistaken we'd only need to consult the value of `MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT` in `ssl_extract_add_data_from_record()` and when defining `MBEDTLS_TLS_EXT_CID`. (And would also make it easier to remove the compat option later.)

Wdyt?

**Review by mpg — CHANGES_REQUESTED** [review]
This is only a first pass, looking at the general structure, which looks pretty good so far!

I'm thinking that perhaps with a different strategy for the compat option, we could reduce the number of places where the dependency would have to change, which would be convenient for testing too.

I found a couple of minor issues, including one that's causing lots of CI failures. I didn't look if there were other sources of failures, I suggest fixing the one first and then seeing what the CI says.

I had a look, and it doesn't seem that OpenSSL or Gnu TLS support this extension, so I don't think we'll be able to add automated interop testing to our CI in the short-to-medium term. Thanks to @boaks for the manual interop testing though!

I'll make a second pass next week, more focused on correctness / security / maintainability of the actual new code, but I already appreciate that everything is isolated in a single, well-commented function, so hopefully everything should go fine!

File: include/mbedtls/mbedtls_config.h [hunk 4]
```diff
@@ -1335,6 +1329,28 @@
...
 //#define MBEDTLS_SSL_DTLS_CONNECTION_ID
 
+/**
+ * \def MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY
+ *
+ * Enable support for the DTLS Connection ID extension utilizing 
+ * the version published in draft-ietf-tls-dtls-connection-id-05,
+ * https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-05
+ *
+ * Support for this functionality is included for backwards-
```
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/mbedtls_config.h:1339]: Note: we first need to revolve the discussion about what options we use, with what meaning: https://github.com/ARMmbed/mbedtls/pull/5061#discussion_r794417327

File: library/ssl_misc.h [hunk 5]
```diff
@@ -289,7 +289,7 @@
    implicit sequence number. */
 #define MBEDTLS_SSL_HEADER_LEN 13
 
-#if !defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
+#if !(defined(MBEDTLS_SSL_DTLS_CONNECTION_ID) || defined(MBEDTLS_SSL_DTLS_CONNECTION_ID_LEGACY))
```
[the remark this answers is not in the document]
Comment (mpg) [reply library/ssl_misc.h:292]: Note: we first need to revolve the discussion about what options we use, with what meaning: https://github.com/ARMmbed/mbedtls/pull/5061#discussion_r794417327

**Comment by boaks** [comment]
> I had a look, and it doesn't seem that OpenSSL or Gnu TLS support this extension, so I don't think we'll be able to add automated interop testing to our CI in the short-to-medium term. Thanks to @boaks for the manual interop testing though!

I added that test to [Eclipse/Californium](https://github.com/eclipse/californium/tree/master/californium-tests/californium-interoperability-tests), see also issue #5443 for a minor improvement.

It's not "fully automated", I generate the c-artifacts manually "on my machine". The Californium build tests are using that to ensure, that the development of Californium itself didn't break that function.

(By the way, unfortunately too many (d)tls libraries are not that useful for IoT. I added a simple dtls cid client in the meantime to [Eclipse/TinyDtls](https://github.com/eclipse/tinydtls)  to have at least a second client implementation.)

File: include/mbedtls/mbedtls_config.h [hunk 6]
```diff
@@ -1335,6 +1329,28 @@
  */
 //#define MBEDTLS_SSL_DTLS_CONNECTION_ID
```
[the remark this answers is not in the document]
Comment (boaks) [reply include/mbedtls/mbedtls_config.h:1330]: FMPOV,  as long as [DTLS-RRC](https://github.com/tlswg/dtls-rrc) isn't complete, I would recommend to add the guidance, that for systems with  "massive message exchanging and significant traffic amplification", an equivalent protection is required.
But not to stress that too much: the main use-case is for me, power saving sensor networks, which naturally have no amplification. And I also don't assume, that in the next future (before that RRC gets completed) such massive deployment will be realistic.

File: include/mbedtls/mbedtls_config.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hannestschofenig) [reply include/mbedtls/mbedtls_config.h:1341]: Regarding the option setting, I think I have to agree with you.

**Review by daverodgman — CHANGES_REQUESTED** [review]
Please add a Changelog

**Comment by boaks** [comment]
I've tested the latest #7762c6e67151d459458e92b69c70b572168cc15e with Californium successfully.

**Comment by hannestschofenig** [comment]
> FMPOV,
> 
> ```
> #if defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
>     unsigned char add_data[23 + MBEDTLS_SSL_CID_IN_LEN_MAX];
> #else
>     unsigned char add_data[13 + 1 + MBEDTLS_SSL_CID_IN_LEN_MAX ];
> #endif
> ```
> 
> must be used for both, `mbedtls_ssl_encrypt_buf` and `mbedtls_ssl_decrypt_buf`.
> 
> My tests have been successful, because I use a rather short cid.

I have changed the add_data structure to add_data[13] for the case when CIDs are not used.

```
#if defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
    unsigned char add_data[23 + MBEDTLS_SSL_CID_IN_LEN_MAX];
#else
    unsigned char add_data[13];
#endif
```

**Comment by boaks** [comment]
@mpg 
@daverodgman 

FMPOV, this PR is "LGTM".

You both have requested changes, so would it be possible for you to spend some time in review the changes?
It's 2 weeks ago, that Hannes updated it towards the "development-branch". I'm afraid, it will be a endless loop, if reviews takes longer.

**Comment by boaks** [comment]
@hannestschofenig 

I hope the reviews will be done soon. Afterwards I would propose, that the commits are squashed and you add the missing "Signed-off-by: Author Name <authoremail@example.com>" to the resulting commit.

Or are there any reasons, not to squash after the reviews?

**Comment by gilles-peskine-arm** [comment]
@boaks What's reviewed is what we merge, including the commit history. We want the commit history to accurately reflect the development and the review, and we want it to be helpful to readers (one thing at a time), so we never squash, and we avoid rebasing during review unless there's a compelling reason (e.g. merge conflict, or commit that doesn't make sense on its own such as forgetting to add a new file).

**Comment by gilles-peskine-arm** [comment]
The DCO check is failing. Please add signoff lines to all commit messages before we can (re)start reviewing.

**Comment by boaks** [comment]
> What's reviewed is what we merge ...

For work, which is done over such a long period, this will hardly work. Nor do I see, that the most are interested in the history of the withdrawn stuff. At least I prefer to focus on the final outcome and the proper documentation of that.

Anyway, this is not my PR, so proceed as you prefer.

**Comment by plskeggs** [comment]
@hannestschofenig -- is it safe to assume that this PR is superseded by https://github.com/Mbed-TLS/mbedtls/pull/6264?

**Comment by hannestschofenig** [comment]
I created the new PR to deal with the Signed-off-by issue. The new PR also doesn't carry the entire commit history because there was a lot of "forwards-and-backwards" in the design.

**Comment by boaks** [comment]
@hannestschofenig 

Thanks a lot for all the work an the big step ahead now!

**Comment by boaks** [comment]
Some news:

For mbedtls there is now a renewed [PR 6264](https://github.com/Mbed-TLS/mbedtls/pull/6264).

Just for the curious ones, which want to have a first impression and collect some experience with DTLS 1.2 CID and a Thingy:91, I prepared a demo-client with [Eclipse/tinydtls](https://github.com/boaks/zephyr-coaps-client).

**Comment by mpg** [comment]
Closing, superseded by https://github.com/Mbed-TLS/mbedtls/pull/6264

**Pull request closed without merging**